File: Day 12/Day 12 Inheritance.c
#include <iostream>
#include <vector>

using namespace std;


class Person{
	protected:
		string firstName;
		string lastName;
		int id;
	public:
		Person(string firstName, string lastName, int identification){
			this->firstName = firstName;
			this->lastName = lastName;
			this->id = identification;
		}
		void printPerson(){
			cout<< "Name: "<< lastName << ", "<< firstName <<"\nID: "<< id << "\n"; 
		}
	
};

class Student :  public Person{
	private:
		vector<int> testScores;  
	public:
        /*	
        *   Class Constructor
        *   
        *   Parameters:
        *   firstName - A string denoting the Person's first name.
        *   lastName - A string denoting the Person's last name.
        *   id - An integer denoting the Person's ID number.
        *   scores - An array of integers denoting the Person's test scores.
        */
        Student(string firstName,string lastName,int idNumber,vector<int> scores) : Person(firstName, lastName,idNumber) {
        this->testScores= scores;
        }
        /*	
        *   Function Name: calculate
        *   Return: A character denoting the grade.
        */
        char calculate()
        {
            int sum=0,avg=0;
            char grade;
            for(int i=0;i<this->testScores.size();i++)
            { 
                sum+=this->testScores[i];
            }
            avg=sum/this->testScores.size();
            if(90<=avg && avg<=100)
            {
                grade='O';
            }
            if(80<=avg && avg<90)
            {
                grade='E';
            }
            if(70<=avg && avg<80)
            {
                grade='A';
            }
            if(55<=avg && avg<70)
            {
                grade='P';
            }
            if(40<=avg && avg<55)
            {
                grade='D';
            }
            if(avg<40)
            {
                grade='T';
            }
        return grade;
        }        
};  

int main() {
	string firstName;
  	string lastName;
	int id;
  	int numScores;
	cin >> firstName >> lastName >> id >> numScores;
  	vector<int> scores;
  	for(int i = 0; i < numScores; i++){
	  	int tmpScore;
	  	cin >> tmpScore;
		scores.push_back(tmpScore);
	}
	Student* s = new Student(firstName, lastName, id, scores);
	s->printPerson();
	cout << "Grade: " << s->calculate() << "\n";
	return 0;
}

File: Day 12/Day 12 Inheritance.js
'use strict';

var _input = '';
var _index = 0;
process.stdin.on('data', (data) => { _input += data; });
process.stdin.on('end', () => {
    _input = _input.split(new RegExp('[ \n]+'));
    main();    
});
function read() { return _input[_index++]; }

/**** Ignore above this line. ****/

class Person {
    constructor(firstName, lastName, identification) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.idNumber = identification;
    }
    
    printPerson() {
        console.log(
            "Name: " + this.lastName + ", " + this.firstName 
            + "\nID: " + this.idNumber
        )
    }
}

class Student extends Person {
 constructor(firstName, lastName, idNumber,scores) { super(firstName, lastName,idNumber); 
 
        this.testScores= scores;
        }
        calculate()
        {
            let sum=0,avg=0;
            let grade='';
            for(let i=0;i<this.testScores.length;i++)
            { 
                sum+=this.testScores[i];
            }
            avg=sum/this.testScores.length;
            if(90<=avg && avg<=100)
            {
                grade='O';
            }
            if(80<=avg && avg<90)
            {
                grade='E';
            }
            if(70<=avg && avg<80)
            {
                grade='A';
            }
            if(55<=avg && avg<70)
            {
                grade='P';
            }
            if(40<=avg && avg<55)
            {
                grade='D';
            }
            if(avg<40)
            {
                grade='T';
            }
        return grade;
        }        
} 


function main() {
    let firstName = read()
    let lastName = read()
    let id = +read()
    let numScores = +read()
    let testScores = new Array(numScores)
    
    for (var i = 0; i < numScores; i++) {
        testScores[i] = +read()  
    }

    let s = new Student(firstName, lastName, id, testScores)
    s.printPerson()
    s.calculate()
    console.log('Grade: ' + s.calculate())
}
